Another try at merging cloud modules

davidhalter opened the pull request:
Let's review it here.

gaudenz committed:
cloudscale: new module cloudscale_server

gaudenz committed:
cloudscale: integration test for cloudscale_server

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+                                                'stopped',
+                                                'absent',
+                                       ]),
+            name                = dict(),
```
I suppose ansible is internally using type='str', but I still think it would be better (more explicit) to mention it here. I feel like an empty dict is a bit weird.

gaudenz replied:
type='str' is the default, following convention over configuration I don't see much point in explicitly stating defaults. But if you feel strongly about this I don't care too much. Let's see what the Ansible devs think about it?

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+
+            # Remove None values, these don't get correctly translated by urlencode
+            if v is None:
+                data.pop(k)
```
IMO use `del data[k]`, it's cleaner (in both places).

gaudenz replied:
I had it with del first and it felt weird to me, but I don't care much. I'll change that.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+    def _transform_state(server):
+        if 'status' in server:
+            server['state'] = server['status']
+            server.pop('status')
```
Just wondering, why do you need to pop the status here? It feels very strange.

gaudenz replied:
because otherwise we would have state and status in the result. This is basically because our API calls the attribute status while the ansible convention is to call it state. So we want to rename the key in the dictionary.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+            return None
+        else:
+            self._module.fail_json(msg='Failure while calling the cloudscale.ch API with DELETE for '
+                                       '"%s": %s' % (api_call, info['body']))
```
I think we should use a method that looks like:

```
    def _request(self, method, api_call, data=None):
        resp, info = fetch_url(self._module,
                               API_URL+api_call,
                               headers = self._auth_header,
                               method='DELETE'
                               data=data)
       if info['status'] == 201:
            return json.loads(resp.read())
        elif info['status'] == 204:
            return None
        else:
            self._module.fail_json(msg='Failure while calling the cloudscale.ch API with %s for '
                                       '"%s": %s' % (method, api_call, info['body']))
```

instead of repeating ourselves. I'm not against also having helper functions like `_delete(...): return self._request(...)`.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+                                       ]),
+            name                = dict(),
+            uuid                = dict(),
+            flavor              = dict(),
```
Would it make sense to add a default flavor and image?

gaudenz replied:
I would perfer to stay close to the API where we don't have defaults either. I could live with a default flavor, but a default image looks like a can of worms.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+  - An valid API token is required for all operations. You can create as many tokens as you like using the cloudscale.ch control panel at U(https://control.cloudscale.ch)
+notes:
+  - Instead of the api_token parameter the CLOUDSCALE_API_TOKEN environment variable can be used.
+  - To create a new server at least the C(name), C(ssh_key), C(image) and C(flavor) options are required.
```
Could you explain quickly what C(name) means? Is that being parsed and replaced?

gaudenz replied:
See http://docs.ansible.com/ansible/dev_guide/developing_modules.html#documenting-your-module it means constant-width and is recommended for parameters in the documentation.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+  - "For details consult the full API documentation: U(https://www.cloudscale.ch/en/api/v1)"
+  - An valid API token is required for all operations. You can create as many tokens as you like using the cloudscale.ch control panel at U(https://control.cloudscale.ch)
+notes:
+  - Instead of the api_token parameter the CLOUDSCALE_API_TOKEN environment variable can be used.
```
Use dots at the end of the sentence everywhere or nowhere (in this list).

gaudenz replied:
OK I went for dots everywhere.

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+        changed = True
+    elif current_state == 'running' and target_state == 'stopped':
+        server.stop_server()
+        changed = True
```
What do we do when the state is changing? Not that I think this is wrong, but I would like to clarify this. Maybe also with a comment in the code.

gaudenz replied:
The current code just does nothing and returns the state as changing. I see a few options:
1) keep it as it is and document this. Seems a bit suboptimal as the task succeeds without ensuring the correct state.
2) just fail. Might be a bit hard, but at least it's correct. We currently can't ensure the right state, so we fail.
3) as in 2) but first wait for one of the other states. So we only fail if we don't reach a 'good' state until the timeout.

What do you think?

This also made me realize that we currently don't fail if the desired state can't be reached. I guess we should change that too and fail with an error message.

gaudenz replied:
And in any case I should readd the explcit current_state == target_state check I had at one point because it's better to be safe than sorry. The server could also be paused or in an error state. Don't know how the API presents that.

davidhalter replied:
I think I would prefer option 3 (if it's not too much of a hassle). The case `error` and `paused` are very seldom and not a state that a servers might be in. `changing` however is very real and happens in a lot of circumstances (and there might be even more in the future).

davidhalter commented on lib/ansible/modules/cloud/cloudscale/cloudscale_server.py:
```diff
@@ -0,0 +1,438 @@
+
+    def start_server(self):
+        self._post('servers/%s/start' % self.info['uuid'])
+        self.wait_for_state('running')
```
When the server is running, it doesn't mean you can immediately SSH on it. Do we need to document that?

gaudenz replied:
I don't think so. We don't do that anywhere else. This method was mostly introduced for two reasons:
1) To have the complete information about the server in the return value
2) To be able to transition from absent to stopped which means creating, waiting for running and then stopping.

gaudenz committed:
Add dots at the end of all sentences

gaudenz committed:
Use del dict[k] instead of dict.pop(k)

gaudenz committed:
Fail if state changes run into a timeout

If the state does not change until the timeout is over. Something
probably went wrong and it's safer to fail.

gaudenz committed:
Remove superflous empty line

gaudenz commented:
I have now updated my branch with the changes incorporated from your comments and discovered by the ansible CI system. The only open point is how to handle unusual states (changing, paused, error).

gaudenz commented:
BTW the pull request to ansible is here:
https://github.com/ansible/ansible/pull/20175

davidhalter commented:
I have read all of your comments and think we're in a good position to have it merged.

I think the only thing that we have not talked about (besides the state thing), is that we have to list the flavors and images somewhere and link to it.

gaudenz committed:
Check for allowed server state at module startup

For this moudle to be able to do anything the server must be in one of
the allowed states 'running', 'stopped' or 'absent'. All other states
are either error states or short lived transition states ('changeing').
In this case wait for the server to reach one of the allowed states or
fail after a timeout.

gaudenz committed:
Change timeout to 30 seconds

The timeout was always meant to be 30 seconds, but was set to 30 days
unintentionally.

davidhalter commented:
I like the last fixes. Let's see what the other reviewers think.
